framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js

```diff
@@ -42,47 +43,199 @@ class BlockSynchronizationMechanism {
+		this.channel.publish('chain:processor:sync', { block: receivedBlock });
+		throw error;
+	}
+
+	/**
+	 * Reverts the current chain so the new tip of the chain corresponds to the
+	 * last common block.
+	 * @param receivedBlock
+	 * @param {Object} peer - The selected peer to target.
+	 * @return {Promise<void>}
+	 * @private
+	 */
+	async _revertToLastCommonBlock(receivedBlock, peer) {
+		const lastCommonBlock = await this._requestLastCommonBlock(peer);
+
+		if (!lastCommonBlock) {
+			// TODO: Should we apply penalty here?
+			this._applyPenaltyAndRestartSync(
+				peer,
+				receivedBlock,
+				new Error(
+					`No common block has been found between the system chain and the targeted peer ${
+						peer.id
+					}`,
+				),
 			);
+		}
 
-			const { valid: validBlock, err } = await this._blockDetachedStatus(
-				networkLastBlock,
+		if (lastCommonBlock.height < this.bft.finalizedHeight) {
+			this._applyPenaltyAndRestartSync(
+				peer,
+				new Error(
+					'The last common block height is less than the finalized height of the current chain',
+				),
+			);
+		}
+
+		try {
+			await deleteBlocksAfterHeight(
+				this.processorModule,
+				this.blocks,
+				lastCommonBlock.height,
 			);
-			const notInDifferentChain = ForkChoiceRule.isDifferentChain(
-				this.blocks.lastBlock,
-				networkLastBlock,
+		} catch (e) {
+			this.logger.debug(
+				`Failure deleting blocks after height ${
+					lastCommonBlock.height
+				} while reverting chain to last common block`,
 			);
+			throw e;
+		}
+	}
+
+	/**
+	 * Requests the last common block in common with the targeted peer.
+	 * In order to do that, sends a set of network calls which include a set of block ids
+	 * corresponding to the first block of descendent consecutive rounds (starting from the last one)
+	 * @param peer - The peer to target.
+	 * @return {Promise<Object | undefined>}
+	 * @private
+	 */
+	async _requestLastCommonBlock(peer) {
+		const [lastBlock] = await this.storage.entities.Block.get({
+			sort: 'height:desc',
+			limit: 1,
+		});
 
-			if (!validBlock || !notInDifferentChain) {
-				this.channel.invoke('network:applyPenalty', {
-					peerId: bestPeer.id,
-					penalty: 100,
-				});
-				this.channel.publish('chain:processor:sync', { block: receivedBlock });
-				throw err;
+		const blocksPerRequestLimit = 10; // Maximum number of block IDs to be included in a single request
+		const requestLimit = 10; // Maximum number of requests to be made to the remote peer
+
+		let numberOfRequests = 0; // Keeps track of the number of requests made to the remote peer
+		let numberOfBlocks = 0; // Keeps track of the number of block IDs to be included in a request
+		let currentRound = Math.floor(
+			lastBlock.height / this.constants.activeDelegates,
+		); // Keeps track of the round number it is  being used to compute the first block of the round to be included in the request payload
+		let highestCommonBlock; // Holds the common block returned by the peer if found.
+
+		while (
```

Comment:
This nested while loop structure looks a bit complex, maybe we can find an easier way to rewrite this? The cyclomatic complexity of this function is really high and needs a lot of comments to understand the idea that's happening here. I would go for a more functional approach (but I understand this is a draft!) 👍

Reply:
I agree, i think it's bit too complex here

Reply:
Moved the inner while to a separate function. Please check again when you have time :)